### PR TITLE
fix for panic in document tokenizer

### PR DESCRIPTION
**Pull request by urabenstein**
Workaround to prevent panics occurring because byte indices do not point to char boundaries.

### CODE REVIEW

**Comment by dginev** [comment]
Hey @urabenstein hope you're still around and seeing notifications! Apologies for not getting to this pull request earlier, I believe I had postponed the pattern matcher work until Frederik is back around.

Thanks for the pull request - I ran into some malformed unicode panics myself just now #10 and will try to incorporate a version of your upgrades.

File: examples/document-tokenizer.rs [hunk 1]
```diff
@@ -221,9 +241,33 @@ fn annotate(node : Node, root: &Node, range: &DNMRange, dnm: &DNM, dom: &DOM) ->
         act_end.unlink();
         node.add_child(&act_end).unwrap();
     }
+
     return true;
 }
 
+/* Takes a possibly invalid byte index to plaintext and returns a close and valid byte index. A byte index is invalid iff it does
+   not point to a char boundary */
+fn fix_byte_index(mut byte_index : usize, plaintext : String) -> usize {
```
Comment (dginev) [inline examples/document-tokenizer.rs]: I think this is the core fix needed for the DNM ranges / tokenization boundaries - they must always point at a char boundary. An approach I may borrow and provide fixes natively in the methods affected.

File: src/dnm.rs [hunk 2]
```diff
@@ -199,7 +200,18 @@ impl <'dnmrange> DNMRange <'dnmrange> {
         trimmed_start = self.start;
         trimmed_end = self.start;
     }
-    DNMRange {start : trimmed_start, end: trimmed_end, dnm: self.dnm}
+
+    //decrease the byte index to the next non-boundary index
```
Comment (dginev) [inline src/dnm.rs]: I think this should be a part of the main loop rather than a final sanity check, then we can avoid some edge cases. Compound unicode characters are a tricky thing in the first place, but the `is_char_boundary` check goes a long way to stay safe.

**Comment by dginev** [comment]
Would have loved to have a test btw, since now I'm a little uncertain if I will patch the correct issues.

**Comment by dginev** [comment]
Fixed via #eb16446 , just doubled down on using `.len_utf8` when doing char arithmetic

**Comment by dginev** [comment]
Thanks for the PR, helped a lot @urabenstein !
